File: framework/languages/dotnet/samples/SupportChat/Server/Support/Infrastructure/ZLink/Actors/SupportUserActorRelocationAdapter.cs
using System.Text.Json;
using Zlink.Framework.Contracts.Actors;

namespace SupportChat.Server.Support.Infrastructure.ZLink.Actors;

internal sealed class SupportUserActorRelocationAdapter
    : IZLinkActorRelocationAdapter<SupportUserActor>
{
    public ValueTask<byte[]> CaptureAsync(
        SupportUserActor actor,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        return ValueTask.FromResult(JsonSerializer.SerializeToUtf8Bytes(
            new SupportUserActorRelocationState(
            actor.DisplayName,
            actor.Role,
            actor.ParticipantId,
            actor.ConversationId,
            actor.CaptureCompletedJoinOperations())));
    }

    public ValueTask RestoreAsync(
        SupportUserActor actor,
        ReadOnlyMemory<byte> payload,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        var transferred = JsonSerializer.Deserialize<SupportUserActorRelocationState>(
            payload.Span) ?? throw new InvalidDataException("Actor relocation state is empty.");
        actor.SetIdentity(transferred.DisplayName, transferred.Role, transferred.ParticipantId);
        if (!string.IsNullOrEmpty(transferred.ConversationId))
            actor.JoinConversation(transferred.ConversationId);
        actor.RestoreCompletedJoinOperations(transferred.CompletedJoinOperations);
        return ValueTask.CompletedTask;
    }

    private sealed record SupportUserActorRelocationState(
        string DisplayName,
        string Role,
        string ParticipantId,
        string ConversationId,
        string[] CompletedJoinOperations);
}

File: framework/languages/java/samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/actors/SupportUserActorRelocationAdapter.java
package systems.zlink.samples.supportchat.server.support.actors;

import com.fasterxml.jackson.databind.ObjectMapper;
import systems.zlink.framework.actors.ZLinkActorRelocationAdapter;
import systems.zlink.framework.actors.ZLinkRelocationCancellation;

public final class SupportUserActorRelocationAdapter
    implements ZLinkActorRelocationAdapter<SupportUserActor> {
    private static final ObjectMapper JSON = new ObjectMapper();

    @Override
    public java.util.concurrent.CompletionStage<byte[]> capture(
        SupportUserActor actor,
        ZLinkRelocationCancellation cancellation) {
        try {
            return java.util.concurrent.CompletableFuture.completedFuture(
                JSON.writeValueAsBytes(new TransferState(
                    actor.displayName(),
                    actor.role(),
                    actor.participantId(),
                    actor.conversationId(),
                    actor.pendingConversationId(),
                    actor.completedJoinOperations())));
        } catch (java.io.IOException error) {
            return java.util.concurrent.CompletableFuture.failedFuture(error);
        }
    }

    @Override
    public java.util.concurrent.CompletionStage<Void> restore(
        SupportUserActor actor,
        byte[] state,
        ZLinkRelocationCancellation cancellation) {
        try {
            TransferState transferred = JSON.readValue(state, TransferState.class);
            actor.setIdentity(
                transferred.displayName(),
                transferred.role(),
                transferred.participantId());
            actor.joinConversation(transferred.conversationId());
            actor.restorePendingConversationJoin(
                transferred.pendingConversationId());
            actor.restoreCompletedJoinOperations(
                transferred.completedJoinOperations());
            return java.util.concurrent.CompletableFuture.completedFuture(null);
        } catch (java.io.IOException error) {
            return java.util.concurrent.CompletableFuture.failedFuture(error);
        }
    }

    public record TransferState(
        String displayName,
        String role,
        String participantId,
        String conversationId,
        String pendingConversationId,
        java.util.Set<systems.zlink.framework.actors.ZLinkActorJoinOperationId>
            completedJoinOperations) {
    }
}
